docs/design/2022-08-04-lazy-constraint-check.md

```diff
@@ -0,0 +1,138 @@
+
+## Design
+
+This feature makes it possible to skip acquiring pessimistic locks for the keys that need uniqueness constraint checks in pessimistic transactions. We do constraint checks for these keys during prewrite.
+
+```sql
+CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY);
+INSERT INTO t1 VALUES (1), (2);
+BEGIN PESSIMISTIC;
+SELECT * FROM t1 WHERE id = 1 FOR UPDATE; -- SELECT FOR UPDATE locks key as usual.
+INSERT INTO t1 VALUES (2); -- Skip acquiring the lock and return success.
+COMMIT; -- ERROR 1062 (23000): Duplicate entry '2' for key 'PRIMARY'
+```
+
+### Applicable Scenarios
+
+When talking about the success rate, we are based on the following assumption:
+
+**The user treats duplicated entry errors returned by DMLs as fatal errors which must abort the transaction.**
+
+A duplicated entry error only makes a single DML statement fail in a pessimistic transaction. The transaction does not abort immediately. If the user can handle the duplicated entry error in their application and continue the transaction, skipping pessimistic locks definitely changes the behavior and lowers the success rate.
+
+While under such an assumption, the general success rate won't change with the feature enabled in most cases.
+
+Suppose that transaction A wants to insert a new row with a unique key, or update a field with a unique key.
+
+1. If a duplicated record exists before the DML, when to check the constraint does not change the success rate. Currently, the duplicated entry error is raised when executing the DML. With lazy constraint check, the error is raised when committing the transaction. 
+
+2. If the duplicated record is written by transaction B between transaction A's executing the DML and committing, let's consider different cases:
+
+   1. If both transactions acquire the lock of the unique key, which is the current behavior, transaction B will be blocked until transaction A commits. After transaction A commits, the DML of transaction B will fail due to the duplicated entry.
+   2. If transaction A skips acquiring the lock while transaction B does not, transaction A will fail unless transaction B aborts. When transaction A prewrites, it will either meet the lock or the write record from transaction B. In either case, transaction A will fail to pass the constraint check.
+   3. If both transaction A and B skip acquiring the lock, it's similar to the case of an optimistic transaction. The transaction that commits first will succeed.
+
+   Therefore, although the time when the error is raised is different, only one of the transactions will fail.
+   
+In a special case, the success rate may drop due to the write conflict check:
+
+```sql
+/* init */ CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY);
+/* init */ INSERT INTO t1 VALUES (1);
+
+/* s1 */ BEGIN PESSIMISTIC;
+/* s2 */ DELETE FROM t1 WHERE id = 1;
+/* s1 */ INSERT INTO t1 VALUES (1);
+/* s1 */ COMMIT;
+```
+
```

Comment:
For such cases, 
```
/* init */ CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY, value int);
/* init */ INSERT INTO t1 VALUES (1, 1);

/* s1 */ BEGIN PESSIMISTIC;
/* s1 */ INSERT INTO t1 VALUES (1, 2) LAZY CHECK; -- Skip acquiring the lock. So the statement would succeed.
/* s1 */ SELECT * FROM t1 FOR UPDATE; -- Here the pessimistic lock on rowkey '1' would be acquired
```
Here it's a bit difficult to decide the execution result of the `for update` statement, should the memory buffer content be returned or the row value(1, 1), or a duplicated entry error detected acquiring the pessimistic lock?

Actually, the snapshot read semantic becomes weird in some cases where the pessimistic current read is used together such as https://github.com/pingcap/tidb/issues/24195. A side effect may be that the current read semantic would become weird too in some special cases, especially some historical technical debt like the behaviors of the `unionScan` executor.

Reply:
Right... This makes the case of #24195 more complex.

I don't think the complete solution to this problem should be part of this proposal. So, for this case, personally I think we can treat it as if the pessimistic lock of the `INSERT` gets lost (due to failed pipelined or in-memory locking). This can happen in a normal pessimistic transaction now. In this case, the value in the table _can_ be different from those in the buffer.

Reply:
> treat it as if the pessimistic lock of the INSERT gets lost

Make sense, so we could just ignore the write result if it's conflicted with some returned current read result.

Reply:
I think "Acquisition of lock always resolves constraint check" is simpler in semantics, which requires 1 additional state of keys in prewrite: `no lock + check`. 
While treating as if the lock is lost requires 2 additional states: `lock + check` and `no lock + check` in prewrite. It also introduces complexity in the read/write procedures, which increases the risk of doing it right.

It seems the cost is throwing "Duplicate key" errors in `SELECT FOR UPDATE` or other DMLs that acquire locks. I assume users should be ready for handling such cases since the feature is not for everyone.

Reply:
Umm, yes. As long as we can accept throwing "duplicated entry" errors in a read-only statement, "Acquisition of lock always resolves constraint check" is the easiest to implement and brings least problems.

Reply:
I follow @ekexium's idea and add a section to define the behavior in this case.

Reply:
> Here it's a bit difficult to decide the execution result of the for update statement, should the memory buffer content be returned or the row value(1, 1), or a duplicated entry error detected acquiring the pessimistic lock?
If 'deferred insert' mixed with 'lock if exists', it will make pessimistic lock invalid in some cases. The feature 'lock if exists' may not coexist with 'deferred insert'.
create table t(a int primary key, b int)
begin;
insert into t values(1,1) /* lazy checked*/ ; -- cached in buffer and don't acquire pessimistic lock or unique check
select * from t where a = 1 for update /*lock if exists*/; -- tikv can't find the record and doesn't make a pessimistic lock
commit

Reply:
> insert into t values(1,1) /* lazy checked*/ ; -- cached in buffer and don't acquire pessimistic lock or unique check
> select * from t where a = 1 for update /_lock if exists_/; -- tikv can't find the record and doesn't make a pessimistic lock
> commit

In this case, the late acquisition of lock should override lock_if_exists. Maybe we have to get from the mem buffer first to decide how we lock the keys.